Implement TRON

abelsiqueira opened the pull request:
Implementation of the algorithm described in "Newton's Method for Large Bound-Constrained Optimization Problems" by Lin, C.-J. and Moré, J. J.
There are CUTEst tests selected from a small set being run, but without expectation of convergence.

dpo commented:
Thanks! I'm going to look in detail. Note that `steihaug` is already implemented in `Krylov.cg`.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,212 @@
+      mem = n
+    end
+    H = LBFGSOperator(n, mem, scaling = true)
+  end
```
I feel like it's up to the model to decide whether it's using a quasi-Newton Hessian. It would be easy to define a `LBFGSModel` or `LSR1Model` from any `AbstractNLPModel`. I think that's the way to go.

abelsiqueira replied:
But if I'm using, e.g., CUTEstModel, should I first create a LBFGSModel from that model, and then pass to tron?
Also how would the BFGS update? If it demands an additional call, it would have to be in the AbstractModel. If stored inside the structure, it would have to automatically update when computing two gradients?

dpo replied:
What I have in mind is something like:
```
model = CUTEstModel(...)
lbfgs_model = LBFGSModel(model)
```
Where `type LBFGSModel <: AbstractNLPModel` and the methods `hess(model :: LBFGSModel, ...)`, `hprod(model :: LBFGSModel, ...)`, etc., override the default `hess()`, `hprod()`, etc.

I'd like to know what you think the best way to implement quasi-Newton updates is. What I've been doing in Python is check whether the model has a specific method. I'm not sure it's the best or the most elegant way.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,212 @@
+
+  x = max(min(nlp.meta.x0, u), l)
+  gx = g(x)
+  ϵ = atol + rtol * norm(gx)
```
epsilon should probably depend on the projected gradient norm?!

abelsiqueira commented:
I will take a look at the Steihaug implementation on Krylov and update the $\epsilon$.

dpo commented:
Did you run any benchmarks? For example, you could compare it to Trunk on unconstrained problems.

abelsiqueira commented:
Running the benchmarks now. It is losing a lot, but I haven't made any upgrades yet.
About the Steihaug vs cg, I'm using a modified version of Steihaug that fixes some variables. This is equivalent to using the Z'*A*Z reduced matrix, where Z = eye(n)[:,J] for some set of index J.

To use Krylov.cg I'm thinking about two approaches:
(i) creating a new AbstractLinearOperator that takes A and J and returns Z'*A*Z.
(ii) creating a new AbstractLinearOperator that takes J and returns Z.
 What do you think? I'm thinking the complexity is the same, but the second one should be simpler and cleaner.

dpo commented:
I think option ii is best. Check https://github.com/PythonOptimizers/pykrylov/blob/develop/pykrylov/linop/linop.py#L560.

dpo commented:
Note also that the original Tron has different trust-region management, which can make a difference.

abelsiqueira commented:
Just confirming, option ii is best, but the link you passed is for option i, right?

dpo commented:
Yes, you're right. With your `ReducedIndexOperator`, you get Z, and then you can define Z' * A * Z.

abelsiqueira commented:
And that is working. I'll remove Steihaug, and see the TR update difference.

About the BFGSModel. The problem with this approach is that all methods will always have to check for this, and that's not straightforward. While an automatic approach might make simple methods, like Cauchy methods, more expensive.
If I have to choose, then the first option.

dpo commented:
https://github.com/JuliaSmoothOptimizers/NLPModels.jl/pull/28 shows what I have in mind. Let me know what you think.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,212 @@
+function active(x, l, u; ϵlu = 1e-4)
+  if isa(ϵlu, Float64)
+    ϵlu = (u-l)*ϵlu
+  end
```
Is it worth creating a new vector for this? Could we just use `(u[i] - l[i]) * ϵlu` in the loop?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,212 @@
+    # Model
+    if !bfgs
+      H = LinearOperator(n, Float64, v -> hprod(nlp, x, v))
+    end
```
This can be a `hess_op()`.

abelsiqueira commented:
Thanks for the comments, I'm updating locally. I've made the change of the trust-region but TRON is still not good. I'm comparing with Trunk, and ROSENBR takes over thousands of iterations to converge.

abelsiqueira commented:
Found it. Sign error.

abelsiqueira commented:
Rebased and made some small improvements. It's still losing, though. [Tests on Travis](https://travis-ci.org/JuliaSmoothOptimizers/Optimize.jl/jobs/162675657#L1336-L1341).
I can see it improving if we use `hess` and the "incomplete Cholesky factorization", or if we use BFGS. Do you see any other improvements to be made?

dpo commented:
Did you really rebase? This is showing changes in 18 files, a lot of which look just like changes in recent prs. It's difficult to assess such large changes. We should try to keep prs more focused.

abelsiqueira commented:
Sorry, now it's rebased.

abelsiqueira commented:
Pushed another commit reducing the number of hprods.

dpo commented:
Any success with this yet? Could you show some numbers on a few bound-constrained problems?

abelsiqueira commented:
I have to fix the travis error first. If you want to push #26 before, I can rebase too.

Here are some tests: https://gist.github.com/abelsiqueira/3d16bd574cd598aaa4934a2454a06f87
Of the 317 unconstrained and bound-constrained problems, TRON found a fisrt-order stationary point for 154.
Of the 139 bound-constrained problems, TRON found a fisrt-order stationary point for 43.

This was done with 100_000 maximum iterations and 30 s maximum elapsed time (though time is not enforced during the iteration).

abelsiqueira commented:
Updated the [gist](https://gist.github.com/abelsiqueira/3d16bd574cd598aaa4934a2454a06f87) with new tests after some corrections. I'm giving it 5 minutes now, so it's taking longer to finish, tomorrow I'll have the full result.

For now, there are 206 problems with up to 4000 variables.
174 of those converged.
For the unconstrained problems, 97 out of 117 converged.
For the bound-constrained, 77 out of 89.

I'm gonna put this as ready to go, then we can discuss other improvements. I think moving to constrained solvers is best now.

abelsiqueira commented:
[gist](https://gist.github.com/abelsiqueira/3d16bd574cd598aaa4934a2454a06f87) updated.
Total: 255 out of 317 converged.
For the unconstrained problems, 135 out of 173 converged.
For the bound-constrained, 120 out of 144.

dpo commented:
Thanks! For reference, here are the results of the Python implementation:
- [unconstrained](https://gist.github.com/452cfd0da64ecb90dc9d3a8ccfc9c640)
- [bound constrained](https://gist.github.com/7a7db46965b7c77b8cf450f0b152a10a)
(the problems are in AMPL so dimensions may not be exactly the same).

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,397 @@
+    end
+  end
+  return A
+end
```
If we drop Julia 0.4 support, is this a good use case of 0.5's new broadcast feature?

abelsiqueira replied:
Do you mean for creating `A`?
Sorry, I still don't see it.

dpo commented:
The first variant performs better, doesn't it? Also try adding types.
On Mon, Oct 24, 2016 at 23:06 Abel Siqueira <notifications@github.com>
wrote:

> *@abelsiqueira* commented on this pull request.
> ------------------------------
>
> In src/solver/tron.jl
> <https://github.com/JuliaSmoothOptimizers/Optimize.jl/pull/10>:
>
> > +export tron
> +
> +function active(x, l, u; rtol::Real = 1e-8, atol::Real = 1e-8)
> +  A = Int[]
> +  n = length(x)
> +  for i = 1:n
> +    if l[i] == x[i] == u[i]
> +      push!(A, i)
> +    elseif x[i] < l[i] + min(rtol * (u[i]-l[i]), atol)
> +      push!(A, i)
> +    elseif x[i] > u[i] -  min(rtol * (u[i]-l[i]), atol)
> +      push!(A, i)
> +    end
> +  end
> +  return A
> +end
>
> Hum, this seems like a good idea, but I'm not being able to improve the
> function.
> See this
> <https://gist.github.com/abelsiqueira/f93a4b4949677088f81d824036f757c7>.
> I tried many variations to decrease the allocation, to no avail.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaSmoothOptimizers/Optimize.jl/pull/10>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACXaA5kL_66xETZzHGAtKg516D021Smks5q3XIkgaJpZM4J5lXF>
> .
>
-- 
Dominique

abelsiqueira commented:
The first variant doesn't do the `tol` part, it's only uses `atol`. The second variant creates `tol` and the third does the computation directly. The first is just saying that broadcast works in some cases.

abelsiqueira commented:
Rebased

abelsiqueira commented:
I'm gonna split this in smaller PRs with the tools first. I would like to rework TRON on top of #35, if that's possible.

dpo commented:
We've been working on a OO Matlab version that we should use as a model. I'll point you to it when it's online.

abelsiqueira commented:
@dpo, these are the [Fortran running times](https://gist.github.com/abelsiqueira/4022c6f5e5e96109e654bcecad504dea), do they look correct?

dpo commented:
The number of failures is surprising. Did you try with the driver that ships with CUTEst (I haven't tested it in a while)? You may need to adjust a thing or two because the TRON version you have doesn't use the explicit Hessian.

abelsiqueira commented:
`tron-cutest.f`, same link, is my modification of the `driver.f` that came with the TRON code. I essentially removed the `dmin*` functions and the function `dsphesd`, which, from my understanding, computes an approximation to the Hessian using matrix-vector products. I'm using `ush` and `ushp` for the Hessians, and using `dsetsp` to compute the compressed format. The error that keeps happening, is that `dsetsp` needs the diagonal to be provided. In some problems, like `HS4`, the second element of the diagonal is always null, so `ushp` doesn't return it.

By the way, the number of function evaluations on the table was wrong. Updated.

abelsiqueira commented:
PS. Checking the CUTEst's tron

dpo commented:
This works for me with the version of TRON that doesn't require H explicitly: https://gist.github.com/7f94ec2f21423bc140e9322e194f107d

abelsiqueira commented:
I've updated it to use that one (that's the one in CUTEst, right?). The [updated results](https://gist.github.com/abelsiqueira/4022c6f5e5e96109e654bcecad504dea) are indeed much better.

dpo commented:
I simplified the program further and updated the gist. It's a modification of the CUTEst driver, in which I removed the explicit Hessian (except that it's still evaluated to set up hprod).

abelsiqueira commented:
Ah, I see. I don't have the other tron version, though.

dpo commented:
You have read access to the private repository.

abelsiqueira commented:
Updated again. I'll use it to compare against.

dpo commented:
Great. Thanks!

abelsiqueira commented:
Results using `f(x) < fmin + |fmin|*1e-1 + 1e-6` comparison 

![perf-prof](https://user-images.githubusercontent.com/1068752/28296459-ca857b6a-6b3e-11e7-8c37-55b9988164c9.png)

Results excludings problems that failed without returning information (`maxtime` on Fortran, because I kill the solver, `NaN` or `alpha too small` on Julia - which I haven't figured out yet).
![perf-prof-x](https://user-images.githubusercontent.com/1068752/28296643-ddb5edd6-6b3f-11e7-8beb-136806571672.png)

Notice that Fortran fails a lot for taking too long and since I kill it, it doesn't have `f` or `#f` information, while Julia has.

Profiles on `#f + #g + #Hp`.

dpo commented:
That's really quite good! The horizontal axis scale is so short we can consider both algorithms basically equivalent.

dpo commented:
What's the `pjgr` column in the table of results? If it's the projected gradient, it should be very small, and it isn't.

abelsiqueira commented:
It is the projected gradient, but I my `awk` was not reading `1D-2` as `1e-2`. However, I checked and saw that my modification of the stopping criteria for Fortran was using the norm of the gradient at the first iteration, instead of the projected gradient. It should `|pjgr| < 1e-6 * |pjgr0| + 1e-8` now. Running the tests again.

dpo commented:
Yes. The Julia version should use that stopping test too.

abelsiqueira commented:
Done. [Updated version](https://gist.github.com/abelsiqueira/4022c6f5e5e96109e654bcecad504dea). Not much has changed, though.
I'll push the new version soon.

dpo commented:
The Fortran solver doesn't print information at each iteration, which makes it more difficult to debug corner cases.

abelsiqueira commented:
@dpo Current version is passing, and ready for review. There are some TODOs, but they are improvements, and I rather wait on more serious issues first.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+              cgtol = 0.1,
+              fmin :: Real = -1e32,
+              timemax :: Real=60.0,
+              mem :: Integer=5,
```
What's `mem`? If it used to relate to the incomplete factorization precondition, it's not used anymore.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+              η₀ :: Real=1e-4,
+              η₁ :: Real=0.25,
+              η₂ :: Real=0.75,
+              Δmax :: Real=1e6
```
Could the trust-region parameters somehow be associated with a trust-region "object" as in Trunk?

abelsiqueira replied:
I think we could create an `AbstractTrustRegion`, leave the basic `TrustRegion`, and create a new `TRONTrustRegion` that changes the `update!` function. What do you think?

dpo replied:
Sounds good to me. I'm curious if it would improve the one in Trunk.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+      α = σ₃
+    else
+      α = max(σ₁, -0.5*slope/γ)
+    end
```
Could this be `α = γ <= 0.0 ? σ₃ : max(σ₁, -0.5*slope/γ)`?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+
+Find the smallest and largest values of `α` such that `x + αd` lies on the
+boundary. `x` is assumed to be feasible. `nbrk` is the number of violated
+bounds in the given direction.
```
"the number of breakpoints from `x` in the direction `d`" would be more accurate I believe.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+  neg = find( (d .< 0) .& (x .> ℓ) )
+
+  nbrk = length(pos) + length(neg)
+  nbrk == 0 && return 0, 0, 0
```
The last two should be floats.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+  brkmax = 0.0
+  if length(pos) > 0
+    @views steps = (u[pos] - x[pos])./d[pos]
+    brkmin = min.(brkmin, minimum(steps))
```
Out of curiosity, is `minimum(brkmin, steps...)` faster?

abelsiqueira replied:
It looks like `min(brkmin, steps...)` uses more memory and is slower.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+    slope = dot(g,s)
+    qs = ¹/₂sᵀHs + slope
+"""
+function compute_Hs_slope_qs!(Hs::Vector, H::Union{Matrix,AbstractLinearOperator},
```
`AbstractMatrix`?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+Computes the projected direction `y = P(x + d) - x`.
+"""
+function project_step!(y::Vector, x::Vector, d::Vector, ℓ::Vector, u::Vector)
+  y .= x + d
```
Does this allocate a vector for `x + d`?

abelsiqueira replied:
It did. Changing to `.+` doesn't.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+
+"""`x, s = projected_line_search!(x, H, g, d, ℓ, u; μ₀ = 1e-2)`
+
+Performs a projected line search, searching for a direction
```
"searching for a step size `t`..."

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+
+    0.5sᵀHs + sᵀg ≦ μ₀sᵀg,
+
+but without leaving the first face found by `x + d`.  Backtracking is
```
while remaining on the same face as `x + d`.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+but without leaving the first face found by `x + d`.  Backtracking is
+performed from t = 1.0. `x` is updated in place.
+"""
+function projected_line_search!(x::Vector, H::Union{Matrix,AbstractLinearOperator}, g::Vector, d::Vector, ℓ::Vector, u::Vector; μ₀::Real = 1e-2)
```
`AbstractMatrix`?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+  brkmin = Inf
+  brkmax = 0.0
+  if length(pos) > 0
+    @views steps = (u[pos] - x[pos])./d[pos]
```
It's a detail but I like the Python style of always surrounding operators by a space (here, around `./`). What do you think?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+    brkmax = max.(brkmax, maximum(steps))
+  end
+  if length(neg) > 0
+    @views steps = (ℓ[neg] - x[neg])./d[neg]
```
Spaces around `./` if you agree.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+                              s::Vector, g::Vector)
+  Hs = H * s
+  slope = dot(g,s)
+  qs = 0.5*dot(s, Hs) + slope
```
Spaces around `*`.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+  A = Int[]
+  n = length(x)
+  for i = 1:n
+    δ = -Inf < ℓ[i] < u[i] < Inf ? min(rtol * (u[i]-ℓ[i]), atol) : atol
```
Spaces around `-` inside the `min`.

abelsiqueira commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,414 @@
+  elseif stalled
+    "stalled"
+  else
+    "unknown"
```
Only after #35

abelsiqueira commented:
I've implemented the trust region changes, and added it to TRON. However, I'm having some problems with nonnegative Pred. If I ignore the Pred sign, the problems proceed as before. Fortran TRON doesn't fail, though, but I haven't found the difference yet.

abelsiqueira commented:
Found the problem. `H = hess_op(nlp, x, temp)` then changed `x` inplace.

dpo commented:
Excellent! Do you have updated benchmarks?

abelsiqueira commented:
Julia vs Fortran
![perf-prof-evals-jl-vs-f](https://user-images.githubusercontent.com/1068752/28431028-2216a892-6d59-11e7-9453-83a57a4b5c6b.png)
New vs Old on evaluations
![perf-prof-evals-new-vs-old](https://user-images.githubusercontent.com/1068752/28431030-221e26bc-6d59-11e7-80f7-97f37eee9abe.png)
New vs Old on time
![perf-prof-time-new-vs-old](https://user-images.githubusercontent.com/1068752/28431029-221a93c6-6d59-11e7-92dd-ff6f566cfb06.png)

dpo commented:
Where do you think the difference comes from in terms of evaluations? Also, would you mind plotting the time profile Fortran vs Julia to have an idea of where we stand?

abelsiqueira commented:
I found an error on the projected newton step, where I was not checking correctly the stopping criteria (which can go either way), and I've found an extra `H * s` on `cauchy` (last commit).

On the Fortran timing, I have many 0 times. Do I ignore these problems, or use some kind of minimum?

dpo commented:
Let's only compare on problems that take more than, say, 0.1 seconds.

dpo commented:
That's not too bad!

dpo commented:
Do you have a table of raw results (similar to the Fortran table)?

abelsiqueira commented:
[Tables here](https://gist.github.com/abelsiqueira/4022c6f5e5e96109e654bcecad504dea).

abelsiqueira commented:
Saved another `hprod`.

New vs old eval
![perf-prof-evals-new-vs-old](https://user-images.githubusercontent.com/1068752/28826138-4bd36ebe-769f-11e7-9db3-e35d68815731.png)
Julia vs Fortran eval
![perf-prof-evals-jl-vs-f](https://user-images.githubusercontent.com/1068752/28826137-4bd08d66-769f-11e7-9881-0977f3f35030.png)
Julia vs Fortran time
![perf-prof-time-jl-vs-f](https://user-images.githubusercontent.com/1068752/28826188-84559618-769f-11e7-935a-3581aedb8163.png)

dpo commented:
Thanks! How come the robustness isn't the same on the time profile?

abelsiqueira commented:
The time profile is restricted to > 0.1s.

dpo commented:
Ah, right. The results look great. Should I have another look through and should we plan to merge soon?

abelsiqueira commented:
PS. Restarted the tests on the failing OSX 0.5 because `curl` failed.

dpo commented:
Could you please rebase this?

abelsiqueira commented:
Rebased

abelsiqueira commented:
And now fixed.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+
+using LinearOperators, NLPModels
+
+export tron, project!, project_step!, active, breakpoints
```
Do we want all that suff exported, or just `tron`? Perhaps the others could be a general utilities submodule?!

abelsiqueira replied:
I think `breakpoints`, `compute_Hs_slope_qs!`, `active`, `project!` and `project_step!` should be exported and moved to another file (`tools.jl`, `auxiliary.jl`, `utils.jl`?).
The functions `projected_line_search!` and `cauchy` are more specific to TRON, I think.
`projected_newton!` seems more general. Maybe another folder?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    H = hess_op!(nlp, xc, temp)
+
+    # Cauchy starts here
+    αC, s = cauchy(x, H, gx, Δ, αC, ℓ, u, μ₀=μ₀, μ₁=μ₁, σ=σ)
```
Are the two above comments necessary?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    αC, s = cauchy(x, H, gx, Δ, αC, ℓ, u, μ₀=μ₀, μ₁=μ₁, σ=σ)
+
+    s, Hs, cgits, cginfo = projected_newton!(x, H, gx, Δ, cgtol, s, ℓ, u, max_cgiter=max_cgiter)
+    slope = dot(gx,s)
```
Sorry for nitpicking. I like the Python-style syntax with a space between arguments.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    s, Hs, cgits, cginfo = projected_newton!(x, H, gx, Δ, cgtol, s, ℓ, u, max_cgiter=max_cgiter)
+    slope = dot(gx,s)
+    qs = 0.5 * dot(s, Hs) + slope
+    fx = f(x)
```
`fx` was already evaluated before the main loop. Should this be moved?

dpo replied:
Unless `projected_newton!` changed `x`?!

abelsiqueira replied:
It does

dpo replied:
Ok! Haven't read everything yet.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+
+    if num_success_iters == 0
+      tr.radius = min(Δ, norm(s))
+    end
```
What's this for?

abelsiqueira replied:
It's in the matlab code and I copied it here. There is something similar on the Fortran code, but it seems to be only for the first iteration
```fortran
c      On the first iteration, adjust the initial step bound

snorm = dnrm2(n,s,1)
if (iter .eq 1) delta = min(delta,snorm)
```

dpo replied:
Sounds good.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    end
+
+    # Update the trust region
+    update!(tr, norm(s))
```
`norm(s)` could already have been computed above.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    iter >= itmax ? "maximum number of iterations" : "maximum elapsed time"
+  elseif πx <= ϵ
+    "first-order stationary"
+  elseif fx == -Inf
```
This seems a bit dramatic. If `fx == -Inf`, won't there have been some kind of error earlier? How about a test of the form `fx ≤ -min(1, abs(f(x0))) / eps(eltype(x))`?

abelsiqueira replied:
This should have been updated, sorry. We have `unbounded = fx < fmin` now, but `fmin` was a fixed value, and I'm gonna change to this.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+from `x` in the direction `d`.
+"""
+function breakpoints(x::Vector, d::Vector, ℓ::Vector, u::Vector)
+  n = length(x)
```
It seems `n` isn't needed in this function.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+  neg = find( (d .< 0) .& (x .> ℓ) )
+
+  nbrk = length(pos) + length(neg)
+  nbrk == 0 && return 0, 0.0, 0.0
```
Should those zeros be `zero(eltype(x))`?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+  return slope, qs
+end
+
+"""`active(x, ℓ, u; rtol = 1e-8, atol = 1e-8)`
```
We tend to use 1.0e-8 because we're thinking Float64, but your functions are more general. Perhaps 1.0e-8 should be `sqrt(eps(eltype(x)))`?!

dpo replied:
Also, all those vectors should be `Real`, shouldn't they?

abelsiqueira replied:
Didn't understand this last part. You mean change `Vector` to `Vector{Real}`?

dpo replied:
Well you don't want complex vectors or vectors of strings. But somehow they should be `AbstractVector`s so you can call the function with views. Perhaps something like
`AbstractVector{Real}` or `AbstractVector{T} where {T<:Real}`?!

dpo commented:
I wonder how Tron does compared to Trunk on unconstrained problems.

abelsiqueira commented:
Comparison TRON vs Trunk (up to 1000 variables)

Evals
![perf-prof-evals-tron-vs-trunk](https://user-images.githubusercontent.com/1068752/28995425-56ace414-79bf-11e7-9886-7df4d662a379.png)

Time (t >= 0.1)
![perf-prof-time-tron-vs-trunk](https://user-images.githubusercontent.com/1068752/28995426-56aceac2-79bf-11e7-88ab-9d821442c2ff.png)

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+function cauchy(x::Vector, H::Union{AbstractMatrix,AbstractLinearOperator}, g::Vector,
+                Δ::Real, α::Real, ℓ::Vector, u::Vector;
+                μ₀::Real = 1e-2, μ₁::Real = 1.0, σ::Real = 10.0)
+  # TODO: Use brkmin to care for g direction
```
What does this mean?

abelsiqueira replied:
This is for when `alpha <= brkmin` and so I can reuse `H * d`. This will look ugly, and in some cases, won't improve the execution, so I think it's better to leave for a future dedicated PR.
I'll improve the TODO, though.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+  s = zeros(n)
+  Hs = zeros(n)
+
+  project_step!(s, x, -α * g, ℓ, u)
```
Perhaps `project_step!()` could take an optional step size that defaults to 1 so we don't have to allocate a new vector for `-α * g` here?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    search = true
+    while search
+      α /= σ
+      project_step!(s, x, -α * g, ℓ, u)
```
Same here.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    αs = α
+    while search && α <= brkmax
+      α *= σ
+      project_step!(s, x, -α * g, ℓ, u)
```
another one

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    end
+    # Recover the last successful step
+    α = αs
+    s = project_step!(s, x, -α * g, ℓ, u)
```
and another one.

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+  exit_pcg = false
+  exit_itmax = false
+  iters = 0
+  project!(x, x .+ s, ℓ, u)
```
Here's another allocation. There would probably be no allocation if we wrote directly
```
x .= max.(ℓ, min.(x .+ s, u))
```
but that would be less general because it wouldn't be possible to supply user-defined projections.

abelsiqueira replied:
I can do `x .= x .+ s` and then project.

dpo replied:
👍

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+
+    # Projected line search
+    # TODO: Don't create new arrays here
+    xfree = x[ifree]
```
`@views`?

dpo commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+    end
+  end
+
+  return s, Hs, iters, status
```
`status` is never set.

dpo commented on src/trust-region/basic-trust-region.jl:
```diff
@@ -0,0 +1,40 @@
+  increase_threshold :: Float64
+  decrease_factor :: Float64
+  increase_factor :: Float64
+  ratio :: Float64
```
Should the ratio really be part of the structure? It's computed externally and isn't really a property of the trust region itself.

abelsiqueira replied:
I made this change because TRON updates the trust region using the ratio. The alternative would be changing `update!` to also receive `ratio`, and change `trunk` accordingly. Another alternative would be not having unified trust region `update!`.

abelsiqueira commented on src/solver/tron.jl:
```diff
@@ -0,0 +1,403 @@
+        search = qs >= μ₀ * slope
+      end
+      # TODO: Correctly assess why this fails
+      if α < 1e-200
```
Specifically, I'm doing `sqrt(nextfloat(zero(alpha)))`.

On the rework, what do you suggest? I just like to use `Real` instead of `Float64` because sometimes I write `foo(x = 1)`, which will work for `x::Real`, but not for `x::Float64`.

abelsiqueira commented:
Other issus fixed.

dpo commented on src/solver/tron.jl:
```diff
@@ -180,7 +180,7 @@ function projected_line_search!{T <: Real}(x::AbstractVector{T},
 
   project_step!(s, x, d, ℓ, u, α)
-  project!(x, x .+ s, ℓ, u)
+  x .= x .+ s
```
Does it need to be projected or not?

abelsiqueira replied:
The matlab code had
```
s = gpstep(x, alpha, w, xl, xu)
x = mid(x + alpha * w, xl, xu)
```
but that means that `s = P(x + alpha * w) - x = xnew - x`, so I simplified. Unless it was done for numerical reasons.

abelsiqueira commented:
This has been updated with a change to `fmin`. When `f(x0)` was zero, it would stop early. This new version is pretty much the same as the one from previous tests. It should be ready to go now.

dpo commented:
Excellent! Many thanks for all the hard work @abelsiqueira !
